packages/tools-iottwinmaker/src/lib/workspace.ts

```diff
@@ -0,0 +1,182 @@
+ * @param uniqueId identifier for the role and policy used as suffix in name
+ * @param roleType Role or DashboardRole
+ * @param roleAssumePolicy stringified JSON of the policy
+ * @param rolePolicy stringified JSON of the role
+ * @returns promise of the roleArn
+ */
+async function createRoleAndPolicy(
+  uniqueId: string,
+  roleType: 'WorkspaceRole' | 'WorkspaceDashboardRole',
+  roleAssumePolicy: string,
+  rolePolicy: string
+) {
+  const roleName = `IoTTwinMaker${roleType}-${uniqueId}`;
+  console.log(`Creating Role ${roleName}...`);
+  const createRoleResponse: CreateRoleCommandOutput = await aws().iam.createRole({
+    RoleName: roleName,
+    Path: '/',
+    AssumeRolePolicyDocument: roleAssumePolicy,
+    Description: `${roleType} generated by AWS IoT TwinMaker Getting Started Script`,
+  });
+
+  if (!createRoleResponse.Role || !createRoleResponse.Role.Arn) {
+    throw new Error(`Failed to create role ${roleName}.`);
+  }
+
+  const roleArn: string = createRoleResponse.Role.Arn;
+  console.log(`Created Role: ${roleArn}.`);
+  const createPolicyResponse: CreatePolicyCommandOutput = await aws().iam.createPolicy({
+    PolicyName: `IoTTwinMaker${roleType}Policy-${uniqueId}`,
+    PolicyDocument: rolePolicy,
+    Description: `${roleType} Policy generated by AWS IoT TwinMaker Getting Started Script`,
+  });
+
+  const policyArn: string | undefined = createPolicyResponse.Policy?.Arn;
+  console.log(`Created Policy: ${policyArn}`);
+  await aws().iam.attachRolePolicy({
+    RoleName: roleName,
+    PolicyArn: policyArn,
+  });
+  console.log(`Attached Policy: ${policyArn} to Role: ${roleArn}.`);
+  return roleArn;
+}
+
+/**
+ * Helper function during workspace creation to create S3 bucket
+ * @param uniqueId identifier for bucket used as suffix in the name
+ * @returns promise of the role S3 bucket name and arn
+ */
+async function createWorkspaceS3Bucket(uniqueId: string) {
+  const s3BucketName = `iot-twinmaker-workspace-bucket-${uniqueId}`;
+  console.log('Creating S3 Bucket for the TwinMaker Workspace...');
+  const createBucketResponse: CreateBucketCommandOutput = await aws().s3.createBucket({ Bucket: s3BucketName });
+  console.log(`Workspace S3 Bucket created: ${createBucketResponse.Location}`);
+  console.log('Configuring CORS policy...');
+  aws().s3.putBucketCors({
+    Bucket: s3BucketName,
+    CORSConfiguration: {
+      CORSRules: [
+        {
+          AllowedHeaders: ['*'],
+          AllowedMethods: ['GET', 'PUT', 'POST', 'DELETE', 'HEAD'],
+          AllowedOrigins: ['*'],
+          ExposeHeaders: ['ETag', 'Access-Control-Allow-Origin'],
+        },
+      ],
+    },
+  });
+  return { s3BucketName, s3BucketArn: `arn:aws:s3:::${s3BucketName}` };
+}
+
+/**
+ * Driver function for workspace creation
+ * @param workspaceId workspace-id to be created
+ * @returns promise of workspace arn, s3 arn, role arn, dashboardrole arn
+ */
+async function prepareWorkspace(workspaceId: string) {
```

Comment:
does the creation logic match what we do in the console? curious if we can reuse some logic there

e.g. I believe in console the created bucket has access logging enabled as well as object versioning?

Reply:
I notice the created bucket has those 2 things disabled, is there a way for my to enable that using the SDK?

I checked this documentation: (https://docs.aws.amazon.com/AmazonS3/latest/userguide/create-bucket-overview.html) and don't seem to see information about this.

In the TS file, our input options include this in the createBucket request:
```
/**
     * <p>The canned ACL to apply to the bucket.</p>
     */
    ACL?: BucketCannedACL | string;
    /**
     * <p>The name of the bucket to create.</p>
     */
    Bucket: string | undefined;
    /**
     * <p>The configuration information for the bucket.</p>
     */
    CreateBucketConfiguration?: CreateBucketConfiguration;
    /**
     * <p>Allows grantee the read, write, read ACP, and write ACP permissions on the
     *          bucket.</p>
     */
    GrantFullControl?: string;
    /**
     * <p>Allows grantee to list the objects in the bucket.</p>
     */
    GrantRead?: string;
    /**
     * <p>Allows grantee to read the bucket ACL.</p>
     */
    GrantReadACP?: string;
    /**
     * <p>Allows grantee to create new objects in the bucket.</p>
     *          <p>For the bucket and object owners of existing objects, also allows deletions and overwrites of those objects.</p>
     */
    GrantWrite?: string;
    /**
     * <p>Allows grantee to write the ACL for the applicable bucket.</p>
     */
    GrantWriteACP?: string;
    /**
     * <p>Specifies whether you want S3 Object Lock to be enabled for the new bucket.</p>
     */
    ObjectLockEnabledForBucket?: boolean;
    /**
     * <p>The container element for object ownership for a bucket's ownership controls.</p>
     *          <p>BucketOwnerPreferred - Objects uploaded to the bucket change ownership to the bucket
     *          owner if the objects are uploaded with the <code>bucket-owner-full-control</code> canned
     *          ACL.</p>
     *          <p>ObjectWriter - The uploading account will own the object if the object is uploaded with
     *          the <code>bucket-owner-full-control</code> canned ACL.</p>
     *          <p>BucketOwnerEnforced - Access control lists (ACLs) are disabled and no longer affect permissions.
     *          The bucket owner automatically owns and has full control over every object in the bucket. The bucket only
     *          accepts PUT requests that don't specify an ACL or bucket owner full control
     *          ACLs, such as the <code>bucket-owner-full-control</code> canned
     *          ACL or an equivalent form of this ACL expressed in the XML format.</p>
     */
    ObjectOwnership?: ObjectOwnership | string;
```

Reply:
it might be a separate API: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketVersioning.html

but would recommend following up with other folks on the team to see how the console does it